### PR TITLE
[SPARK-18134][SQL] Orderable MapType

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

We can make MapType orderable, and thus usable in aggregates and joins.

Closes #15970

## How was this patch tested?

Unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -663,6 +663,18 @@ class CodegenContext {
           }
         """
       s"${addNewFunction(compareFunc, funcCode)}($c1, $c2)"
+    case map: MapType =>
+      val compareFunc = freshName("compareMap")
+      val compare = this.addReferenceObj("compare", map.interpretedOrdering,
+        classOf[Ordering[MapData]].getCanonicalName)
```
Comment (jinxing64) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: TODO: we should code generate this.

**Comment by jinxing64** [comment]
It seems https://github.com/apache/spark/pull/15970 is not being worked.
I resolved conflicts and add some tests in this pr.

**Comment by hvanhovell** [comment]
@jinxing64 thanks for taking over. I have glanced over the PR, and I miss the explicit sorting of maps. We can't assume that maps are sorted out of the box, for example the following example should evaluate to `true` but it won't in your PR: `map(1, 2, 3, 4) = map(3, 4, 1, 2)`.

**Comment by tejasapatil** [comment]
@hvanhovell : based on [your comment over the jira](https://issues.apache.org/jira/browse/SPARK-18134?focusedCommentId=15693519&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15693519), it seemed that the approach to be used is yet to be finalised. Are we moving ahead with this approach ?

**Comment by jinxing64** [comment]
@hvanhovell
Thanks a lot for comment.
I got you point. I will refine soon.

**Comment by jinxing64** [comment]
Jenkins, retest this plesase.

**Comment by jinxing64** [comment]
It seems the failed SparkR unit tests is not related.
In current change, I added `trait OrderSpecified`, expressions(`BinaryComparison`, `Max`, `Min`, `SortArray`, `SortOrder`) using ordering of data type extends `OrderSpecified`. During analyzing, expressions with map type output will be wrapped in `OrderMaps` , thus changed to be an ordered map.
Expressions with map type output in `Aggregate`, `Distinct` and `SetOperation` will also be wrapped in `OrderMaps`.

**Comment by kiszk** [comment]
Jenkins, retest this please

**Comment by jinxing64** [comment]
Conflicts resolved.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/MapType.scala [hunk 2]
```diff
@@ -33,9 +36,10 @@ import org.apache.spark.annotation.InterfaceStability
  */
 @InterfaceStability.Stable
 case class MapType(
-  keyType: DataType,
-  valueType: DataType,
-  valueContainsNull: Boolean) extends DataType {
+    keyType: DataType,
+    valueType: DataType,
+    valueContainsNull: Boolean,
+    ordered: Boolean = false) extends DataType {
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/MapType.scala]: Do we need to take care of `ordered` in `buildFormattedString` and `jsonValue`?

**Comment by jinxing64** [comment]
@kiszk 
Thanks a lot for comments. Tests passed now. In current change `ordered` is included in `jsonValue`. But I'm not sure it is appropriate.

Thanks again for taking time looking into this :)

**Comment by kiszk** [comment]
@jinxing64 I would like to hear opinions from other reviewers whether `ordered` should be included in these methods like `ToString`, `buildFormattedString`, and `jsonValue`.
IMO, I think we should make these methods consistent, `ordered` is included or not. The method prints the status of `MapType`

**Comment by jinxing64** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
Any update?

**Comment by jinxing64** [comment]
@maropu @kiszk 
In current change, `ordered` is excluded from `toString`, `buildFormattedString`, `jsonValue`; I prefer to keep `ordered` internal and used only when ordering.
Actually this change has been working in my cluster for a while.
Do you agree on the approach -- inject `OrderMaps` when needed ?
I'm happy to keep working on this, please comment.

**Comment by maropu** [comment]
Aha, I think we can't touch the stable interface (`MapType`) until the next release (v3.0). So, I think we need to hold this for a while. Or, we need to look for other approaches to fix this.

**Comment by xxzzycq** [comment]
Does the community currently have a join and group by code that supports mapType?

**Comment by jinxing64** [comment]
@xxzzycq 
Currently no

**Comment by maropu** [comment]
I think we can resume this pr for v3.0. Are u still there and can you resume this? @jinxing64 
Also, can you add `Closes #15970` in the PR description?

**Comment by jinxing64** [comment]
@maropu 
Thanks, and yes I'm still here and I can keep going if this pr is interested. 
I will update this pr this weekend.

**Comment by maropu** [comment]
Thanks!

**Pull request closed without merging**